### PR TITLE
v2 rewrite

### PR DESCRIPTION
**Pull request by aedwardstx**
This is a rewrite of hier_config that intends to accomplish the following:
- Add type hinting
- Pass pylint, pep8, black, and mypy
- Convert tests to pytest
- Increase performance
- Integrate several previously monkey patched methods I had developed e.g. `difference()`

This is only compatible with Python >=3.7

Tasks
- [x] make mypy pass all checks
- [x] make pylint pass all checks
- [x] make black pass all checks
- [x] make pytest pass all checks
- [x] get the travis CI builds working

### CODE REVIEW

File: hier_config/host.py [hunk 1]
```diff
@@ -33,152 +36,116 @@ class Host:
...
+        self._hconfig_tags: List[dict] = list()
+        self._running_config: Optional[HConfig] = None
+        self._generated_config = None
         self._remediation_config = None
-        self.facts = dict()
+        self.facts: dict = dict()
 
     def __repr__(self):
-        return 'Host(hostname={})'.format(self.hostname)
+        return "Host(hostname={})".format(self.hostname)
```
Comment (jtdub) [inline hier_config/host.py:52]: Does it make sense to use an f-string here?

File: hier_config/host.py [hunk 2]
```diff
@@ -15,222 +18,146 @@ class Host:
         import yaml
         from hier_config.host import Host
 
-        options = yaml.load(open('./tests/files/test_options_ios.yml'))
+        options = yaml.load(open('./tests/fixtures/options_ios.yml'), loader=yaml.SafeLoader())
```
Comment (jtdub) [inline hier_config/host.py:21]: What is we had a 'options_file' option that called a method to load the file in yaml and set the returned output to self.hconfig_options?

**Review by jtdub — DISMISSED** [review]
I'll leave it up to you whether to take action on my comments. None of them are critical. Great work on this!

File: hier_config/host.py (hunk 2 above)
Comment (aedwardstx) [reply hier_config/host.py:21]: I tried this but ran into typing issues. I can re-attempt at some point in the future.

**Review by jtdub — APPROVED** [review]

**Pull request merged**